quickstart/aspnet/ContosoTeamStats/RedisConnection.cs

```diff
@@ -0,0 +1,216 @@
+﻿using StackExchange.Redis;
+using System;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace ContosoTeamStats
+{
+    public class RedisConnection : IDisposable
+    {
+        private long _lastReconnectTicks = DateTimeOffset.MinValue.UtcTicks;
+        private DateTimeOffset _firstErrorTime = DateTimeOffset.MinValue;
+        private DateTimeOffset _previousErrorTime = DateTimeOffset.MinValue;
+
+        // StackExchange.Redis will also be trying to reconnect internally,
+        // so limit how often we recreate the ConnectionMultiplexer instance
+        // in an attempt to reconnect
+        private TimeSpan _reconnectMinInterval = TimeSpan.FromSeconds(60);
+
+        // If errors occur for longer than this threshold, StackExchange.Redis
+        // may be failing to reconnect internally, so we'll recreate the
+        // ConnectionMultiplexer instance
+        private TimeSpan _reconnectErrorThreshold = TimeSpan.FromSeconds(30);
+        private TimeSpan _restartConnectionTimeout = TimeSpan.FromSeconds(15);
+        private const int _retryMaxAttempts = 5;
+
+        private SemaphoreSlim _reconnectSemaphore = new SemaphoreSlim(initialCount: 1, maxCount: 1);
+        private SemaphoreSlim _initSemaphore = new SemaphoreSlim(initialCount: 1, maxCount: 1);
+        private ConnectionMultiplexer _connection;
+        private IDatabase _database;
+        private bool _didInitialize = false;
+
+        public async Task InitializeAsync()
+        {
+            if (!_didInitialize)
+            {
+                _didInitialize = true;
+                _connection = await CreateConnectionAsync();
+                _database = _connection.GetDatabase();
+            }
```

Comment:
Need to wrap this in a semaphore, rather than a lower level (to control access to `_isInitialized`)